lib/dynamic_config/datastore_cache.rb

```diff
@@ -1,112 +1,104 @@
+require 'cdo/shared_cache'
 require 'honeybadger/ruby'
 
-# A caching layer that sits in front of a datastore that
-# implements get and set
-
+# A multilayered caching interface intended for use by Cdo::DynamicConfig. Ties
+# together three different components:
+#
+#   1. The long-lived datastore, which in production is DynamoDB. This is the
+#      ultimate source of truth for the current state of our configuration
+#      settings, and will only be read from on initialization and when a
+#      configuration value is updated.
+#
+#   2. The shared cache, which in production is ElastiCache. This is a middle
+#      ground shared between the frontend processes which will be kept up to
+#      date with the content in the datastore and from which processes will
+#      fetch updates on a configurable schedule.
+#
+#   3. The local in-memory cache, which in all environments is a simple hash.
+#      Each frontend processes will reference its local cache to serve the
+#      vast majority of user requests, and will occasionally update its
+#      contents from the shared cache.
 class DatastoreCache
+  SHARED_CACHE_NAMESPACE = "DynamicConfigData".freeze
+
   # @param datastore [Object] a datastore adapter
+  # @param identifier [String] a unique identifier for this instance of
+  #        datastore cache; used to prevent DCDO and Gatekeeper from trying to
+  #        store things in the same place.
   # @param cache_expiration [int] seconds after which a cached entry expires
-  def initialize(datastore, cache_expiration: 30)
-    @cache = {}
+  def initialize(datastore, identifier, cache_expiration: 30)
     @datastore = datastore
-    @cache_expiration = cache_expiration
+    @identifier = identifier
 
-    # A list of change listeners.
-    @listeners = []
+    # Set up our local cache; a simple in-memory hashed used to rapidly respond
+    # to user requests which is occasionally refreshed from the shared cache to
+    # pick up changes.
+    @local_cache = {}
+    @local_cache_expiration = cache_expiration
+    @local_cache_last_refreshed_at = nil
 
-    # Note we intentionally do this before spawning the background thread
-    # to make sure the cache is seeded successfully on init
-    update_cache
-    @update_thread = spawn_update_thread
+    populate_shared_cache
```

Comment:
Do I read this correctly as implying we overwrite the shared cache (=elasticache) every time we start up a local dev instance? I might not be following this correctly, but if I did read it right... that doesn't seem inherently bad if it simplifies things, but are there implications from lots of people flashing the cache at once (say when their dev instances start up)? Is this how we keep the shared cache in sync with the dynamodb "durable record"?

Reply:
Dev instances won’t be using the Elasticache/DynamoDB adapter to store dynamic configuration settings (they use the Memory adapter). This logic, I think, primarily applies to auto scaled / load balanced environments (currently only production) where new web application servers are frequently being launched and we want those servers to get the very latest dynamic configuration settings, so in those infrequent cases we refresh all the items in ElastiCache from DynamoDB.

Reply:
It is definitely the case that this total refresh of the cache is probably unnecessary, but yes as Suresh points out it's really only relevant in production and scaling events are infrequent enough that I'm not expecting the performance implications to be noticeable. If that expectation turns out to be wrong, there are some potential efficiently improvements we could apply at a cost of slightly increased complexity.